workspaces/diff-engine/src/commands/shape.rs

```diff
@@ -0,0 +1,770 @@
+use super::{EndpointCommand, SpecCommand, SpecCommandError};
+use crate::events::shape as shape_events;
+use crate::events::ShapeEvent;
+use crate::projections::ShapeProjection;
+use crate::state::shape::{
+  FieldId, FieldShapeDescriptor, ParameterShapeDescriptor, ProviderDescriptor, ShapeId, ShapeKind,
+  ShapeParameterId, ShapeParametersDescriptor,
+};
+use cqrs_core::AggregateCommand;
+use serde::Deserialize;
+use uuid::Uuid;
+
+#[derive(Deserialize, Debug, Clone)]
+pub enum ShapeCommand {
+  AddShape(AddShape),
+  SetBaseShape(SetBaseShape),
+  RenameShape(RenameShape),
+  RemoveShape(RemoveShape),
+
+  // Shape parameters
+  AddShapeParameter(AddShapeParameter),
+  RemoveShapeParameter(RemoveShapeParameter),
+  RenameShapeParameter(RenameShapeParameter),
+  SetParameterShape(SetParameterShape),
+
+  // Fields
+  AddField(AddField),
+  RenameField(RenameField),
+  RemoveField(RemoveField),
+  SetFieldShape(SetFieldShape),
+}
+
+impl ShapeCommand {
+  pub fn add_shape(base_shape_id: ShapeId, name: String) -> Self {
+    Self::AddShape(AddShape::new(base_shape_id, name))
+  }
+}
+
+#[derive(Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct AddShape {
+  pub shape_id: ShapeId,
+  pub base_shape_id: ShapeId,
+  pub name: String,
+}
+
+impl AddShape {
```

Comment:
Is this for testing?

Reply:
Nope, used in handling of `AddPathParameter` command to add a shape for the new path parameter. Instead of creating an `Event` straightaway, I've opted to create command first and convert that to an event. That way  all the `From<Command> for Event` implementations remain the single source of truth on how stuff is mapped. Will especially come in handy once we start working with command + event contexts as well.

Reply:
ok that's helpful. I think it may change slightly if we make the id generator change